src/collect/mod.rs

```diff
@@ -0,0 +1,252 @@
+//! Trace collectors.
+
+use crate::{errors::HWTracerError, Trace};
+use core::arch::x86_64::__cpuid_count;
+use libc::size_t;
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+#[cfg(collector_perf)]
+pub(crate) mod perf;
+#[cfg(collector_perf)]
+pub(crate) use perf::PerfTraceCollector;
+
+const PERF_DFLT_DATA_BUFSIZE: size_t = 64;
+const PERF_DFLT_AUX_BUFSIZE: size_t = 1024;
+const PERF_DFLT_INITIAL_TRACE_BUFSIZE: size_t = 1024 * 1024; // 1MiB
+
+/// The interface offered by all trace collectors.
+pub trait TraceCollector: Send + Sync {
+    /// Return a `ThreadTraceCollector` for tracing the current thread.
+    fn thread_collector(&self) -> Box<dyn ThreadTraceCollector>;
+}
+
+pub trait ThreadTraceCollector {
+    /// Start recording a trace.
+    ///
+    /// Tracing continues until [stop_collector](trait.ThreadTraceCollector.html#method.stop_collector) is called.
+    fn start_collector(&mut self) -> Result<(), HWTracerError>;
+    /// Turns off the tracer.
+    ///
+    /// [start_collector](trait.ThreadTraceCollector.html#method.start_collector) must have been called prior.
+    fn stop_collector(&mut self) -> Result<Box<dyn Trace>, HWTracerError>;
+}
+
+/// Kinds of collector that hwtracer supports (in order of "auto-selection preference").
+#[derive(Debug, EnumIter)]
+pub enum TraceCollectorKind {
+    /// The `perf` subsystem, as found on Linux.
+    Perf,
+}
+
+impl TraceCollectorKind {
+    // Finds a suitable `TraceCollectorKind` for the current hardware/OS.
+    fn default_for_platform() -> Result<Self, ()> {
+        for kind in TraceCollectorKind::iter() {
+            if Self::match_platform(&kind).is_ok() {
+                return Ok(kind);
+            }
+        }
+        Err(()) // No suitable collector found for this system.
+    }
+
+    /// Returns `Ok` if the this collector is appropriate for the current platform.
+    fn match_platform(&self) -> Result<(), HWTracerError> {
```

Comment:
Probably also `Option`.

Reply:
Not sure about this one. The error conveys information in the failure that `None` couldn't.

Reply:
Yes you're right.